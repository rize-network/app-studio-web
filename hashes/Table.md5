0afa01abeb0235734839e0191664a7a3