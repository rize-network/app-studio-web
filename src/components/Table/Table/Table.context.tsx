import React, { CSSProperties } from 'react';
export type TableViewStyles = {
// Defines a type 'TableViewStyles' which will be used to style components of a table view. This ensures that these styles can be standardized and applied throughout the app.
  table?: CSSProperties;
  thead?: CSSProperties;
  tfoot?: CSSProperties;
  tbody?: CSSProperties;
  tr?: CSSProperties;
  td?: CSSProperties;
  th?: CSSProperties;
  caption?: CSSProperties;
};
const defaultStyles: TableViewStyles = {};
// Initializes an empty 'defaultStyles' object that will be used as a fallback if no custom styles are provided.
const TableStylesContext = React.createContext<TableViewStyles>(defaultStyles);
// Creates a new Context object for 'TableViewStyles' with the default value set to 'defaultStyles'. This allows the styles to be passed down through the component tree without having to pass props manually.
export const TableStylesProvider: React.FC<{
// Exports 'TableStylesProvider', a functional component which wraps its children with the 'TableStylesContext.Provider'. This Provider will allow nested components to access the style configuration.
  children: React.ReactNode;
  styles?: TableViewStyles;
}> = ({ children, styles = defaultStyles }) => (
// The 'styles' prop is optional and defaults to 'defaultStyles' if not provided. When used, 'styles' are propagated down the component hierarchy.
  <TableStylesContext.Provider value={styles}>
    {children}
  </TableStylesContext.Provider>
// Exports a custom hook 'useTableStyles' that encapsulates the logic for consuming the 'TableStylesContext' and returns its value, which are the current style configurations.
);
export const useTableStyles = (): TableViewStyles =>
  React.useContext(TableStylesContext);
