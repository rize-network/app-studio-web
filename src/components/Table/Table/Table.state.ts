import React from 'react';

export const useTableState = () => {
  const [isHovered, setIsHovered] = React.useState(false);
  return { isHovered, setIsHovered };
};
