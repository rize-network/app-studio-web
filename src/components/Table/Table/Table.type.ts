import { ViewProps } from 'app-studio';

export type TableViewStyles = {
  table?: ViewProps;
  thead?: ViewProps;
  tfoot?: ViewProps;
  tbody?: ViewProps;
  tr?: ViewProps;
  td?: ViewProps;
  th?: ViewProps;
  caption?: ViewProps;
};

export interface Column {
  title: string;
  field: string;
}
export interface FooterCell {
  value: string;
  props?: ViewProps;
}
