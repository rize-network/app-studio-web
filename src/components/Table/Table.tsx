import React from 'react';
import {
  TableContainer,
  TableHead,
  TableBody,
  TableRow,
  TableCell,
  TableFooter,
  TableCaption,
  TableView,
  TableHeadCell,
} from './Table/Table.view';
import { TableProvider } from './Table/Table.context';
import { TableLayoutProps } from './Table/Table.props';

export const Table = ({ children, views, onClick }: TableLayoutProps) => {
  return (
    <TableProvider views={views} onRowClick={onClick}>
      {children}
    </TableProvider>
  );
};

Table.Head = TableHead;
Table.HeadCell = TableHeadCell;
Table.Body = TableBody;
Table.Row = TableRow;
Table.Cell = TableCell;
Table.Footer = TableFooter;
Table.Caption = TableCaption;
Table.Container = TableContainer;
Table.Template = TableView;
