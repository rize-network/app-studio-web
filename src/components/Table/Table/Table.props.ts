import React from 'react';
import { Column, FooterCell, TableViewStyles } from './Table.type';
export interface TableProps {
  data: any[];
  columns: Column[];
  footer?: FooterCell[];
  caption?: React.ReactNode;
  styles?: TableViewStyles;
}
export interface TableViewProps extends TableProps {}
export interface TableLayoutProps {
  Head?: React.FC<any>;
  Body?: React.FC<any>;
  Row?: React.FC<any>;
  Cell?: React.FC<any>;
  Footer?: React.FC<any>;
  Caption?: React.FC<any>;
  TableContainer?: React.FC<any>;
  children?: React.ReactNode;
  styles?: TableViewStyles;
}
