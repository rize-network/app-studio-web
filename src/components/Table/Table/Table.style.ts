/**
 * Table Styles
 *
 * Defines the styles for the Table component following the design guidelines:
 * - Typography: Inter/Geist font, specific sizes/weights
 * - Spacing: 4px grid system
 * - Colors: Neutral palette with semantic colors
 * - Rounded corners: Consistent border radius
 * - Transitions: Subtle animations
 */

import { ViewProps } from 'app-studio';
import { TableViewStyles } from './Table.type';

/**
 * Default table styles following the design system
 */
export const DefaultTableStyles: TableViewStyles = {
  table: {
    width: '100%',
    fontFamily: 'Inter, -apple-system, BlinkMacSystemFont, sans-serif',
    borderRadius: '8px', // 2 × 4px grid
    overflow: 'hidden',
    boxShadow: '0 1px 3px rgba(0, 0, 0, 0.1)',
    border: '1px solid',
    borderColor: 'color.gray.200',
    transition: 'all 0.2s ease',
  },
  thead: {
    backgroundColor: 'color.gray.50',
    borderBottom: '1px solid',
    borderBottomColor: 'color.gray.200',
  },
  th: {
    padding: '12px 16px', // 3 × 4px and 4 × 4px grid
    fontWeight: '600',
    fontSize: '14px',
    color: 'color.gray.700',
    textAlign: 'left',
  },
  td: {
    padding: '12px 16px', // 3 × 4px and 4 × 4px grid
    fontSize: '14px',
    borderBottom: '1px solid',
    borderBottomColor: 'color.gray.100',
  },
  tr: {
    transition: 'background-color 0.2s ease',
    _hover: {
      backgroundColor: 'color.gray.50',
    },
  },
  tfoot: {
    backgroundColor: 'color.gray.50',
    fontWeight: '600',
    borderTop: '2px solid',
    borderTopColor: 'color.gray.200',
  },
  caption: {
    margin: '8px 0', // 2 × 4px grid
    color: 'color.gray.600',
    fontSize: '14px',
    fontStyle: 'italic',
  },
};

/**
 * Compact table styles
 */
export const CompactTableStyles: TableViewStyles = {
  ...DefaultTableStyles,
  th: {
    ...DefaultTableStyles.th,
    padding: '8px 12px', // 2 × 4px and 3 × 4px grid
    fontSize: '12px',
  },
  td: {
    ...DefaultTableStyles.td,
    padding: '8px 12px', // 2 × 4px and 3 × 4px grid
    fontSize: '12px',
  },
};

/**
 * Striped table styles
 */
export const StripedTableStyles: TableViewStyles = {
  ...DefaultTableStyles,
  tr: {
    ...DefaultTableStyles.tr,
    _even: {
      backgroundColor: 'color.gray.50',
    },
  },
};

/**
 * Bordered table styles
 */
export const BorderedTableStyles: TableViewStyles = {
  ...DefaultTableStyles,
  td: {
    ...DefaultTableStyles.td,
    borderRight: '1px solid',
    borderRightColor: 'color.gray.100',
  },
  th: {
    ...DefaultTableStyles.th,
    borderRight: '1px solid',
    borderRightColor: 'color.gray.200',
  },
};

/**
 * Dark theme table styles
 */
export const DarkTableStyles: TableViewStyles = {
  table: {
    ...DefaultTableStyles.table,
    backgroundColor: 'color.gray.800',
    borderColor: 'color.gray.700',
  },
  thead: {
    backgroundColor: 'color.gray.900',
    borderBottomColor: 'color.gray.700',
  },
  th: {
    ...DefaultTableStyles.th,
    color: 'color.gray.100',
  },
  td: {
    ...DefaultTableStyles.td,
    color: 'color.gray.300',
    borderBottomColor: 'color.gray.700',
  },
  tr: {
    ...DefaultTableStyles.tr,
    _hover: {
      backgroundColor: 'color.gray.700',
    },
  },
  tfoot: {
    ...DefaultTableStyles.tfoot,
    backgroundColor: 'color.gray.900',
    borderTopColor: 'color.gray.700',
  },
  caption: {
    ...DefaultTableStyles.caption,
    color: 'color.gray.400',
  },
};
