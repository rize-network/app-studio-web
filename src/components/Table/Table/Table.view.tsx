import React from 'react';
import { Element } from 'app-studio';
import { TableViewProps } from './Table.props';
import { useTableStyles } from './Table.context';

export const TableContainer: React.FC<any> = (props) => {
  const styles = useTableStyles();
  return (
    <Element
      as="table"
      borderCollapse="collapse"
      {...styles.table}
      {...props}
    />
  );
};

export const TableHead: React.FC<any> = (props) => {
  const styles = useTableStyles();
  return (
    <Element
      as="thead"
      borderBottom="0.5px solid #ddd"
      textAlign="left"
      color="color.gray.400"
      {...styles.thead}
      {...props}
    />
  );
};

export const TableHeadCell: React.FC<any> = (props) => {
  const styles = useTableStyles();
  return (
    <Element
      as="th"
      padding="14px"
      whiteSpace="nowrap"
      fontWeight="500"
      {...styles.th}
      {...props}
    />
  );
};

export const TableRow: React.FC<any> = (props) => {
  const styles = useTableStyles();
  return <Element as="tr" {...styles.tr} {...props} />;
};

export const TableCell: React.FC<any> = (props) => {
  const styles = useTableStyles();
  return (
    <Element
      as="td"
      padding="14px"
      whiteSpace="nowrap"
      fontWeight={props.isFirstColumn ? '400' : '300'}
      {...styles.td}
      {...props}
    />
  );
};

export const TableBody: React.FC<any> = (props) => {
  const styles = useTableStyles();
  return <Element as="tbody" {...styles.tbody} {...props} />;
};

export const TableFooter: React.FC<any> = (props) => {
  const styles = useTableStyles();
  return <Element as="tfoot" {...styles.tfoot} {...props} />;
};

export const TableCaption: React.FC<any> = (props) => {
  const styles = useTableStyles();
  return (
    <Element
      as="caption"
      margin={'10px 0'}
      color="color.gray.400"
      {...styles.caption}
      {...props}
    />
  );
};

export const TableView: React.FC<TableViewProps> = ({
  data,
  columns,
  footer,
  caption,
}) => {
  return (
    <TableContainer role="Table">
      {caption && <TableCaption>{caption}</TableCaption>}
      <TableHead>
        <TableRow>
          {columns.map((column) => (
            <TableHeadCell key={column.field}>{column.title}</TableHeadCell>
          ))}
        </TableRow>
      </TableHead>
      <TableBody>
        {data.map((row, index) => (
          <TableRow key={index}>
            {columns.map((column, columnIndex) => (
              <TableCell key={column.field} isFirstColumn={columnIndex === 0}>
                {row[column.field]}
              </TableCell>
            ))}
          </TableRow>
        ))}
      </TableBody>
      {footer && (
        <TableFooter>
          <TableRow>
            {footer.map((cell, index) => (
              <TableCell key={index} {...cell.props}>
                {cell.value}
              </TableCell>
            ))}
          </TableRow>
        </TableFooter>
      )}
    </TableContainer>
  );
};
