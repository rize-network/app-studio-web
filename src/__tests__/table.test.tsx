import React from 'react';
import { Table } from '../components/Table/Table';
import renderer from 'react-test-renderer';
import { render, screen, within } from '@testing-library/react';

const columns = [
  { title: 'Invoice', field: 'invoice' },
  { title: 'Payment Status', field: 'paymentStatus' },
  { title: 'Total Amount', field: 'totalAmount' },
  { title: 'Payment Method', field: 'paymentMethod' },
];

const data = [
  {
    invoice: 'INV001',
    paymentStatus: 'Paid',
    totalAmount: '$250.00',
    paymentMethod: 'Credit Card',
  },
  // Add more data as needed for testing
];

describe('Table Component', () => {
  test('renders data ', () => {
    render(
      <Table>
        <Table.Template columns={columns} data={data} />
      </Table>
    );
    expect(screen.getByText('INV001')).toBeInTheDocument();
    expect(screen.getByText('Invoice')).toBeInTheDocument();
  });

  test('renders column headers and data in Table.Template', async () => {
    const { getByRole } = render(
      <Table>
        <Table.Template columns={columns} data={data} />
      </Table>
    );

    const table = getByRole('Table');
    const withinTable = within(table);

    columns.forEach((column) => {
      expect(withinTable.getByText(column.title)).toBeInTheDocument();
    });

    data.forEach((row) => {
      expect(withinTable.getByText(row.invoice)).toBeInTheDocument();
      expect(withinTable.getByText(row.paymentStatus)).toBeInTheDocument();
      expect(withinTable.getByText(row.totalAmount)).toBeInTheDocument();
      expect(withinTable.getByText(row.paymentMethod)).toBeInTheDocument();
    });
  });
});

test('Table to match snapshot', () => {
  const tree = renderer
    .create(
      <Table>
        <Table.Template columns={columns} data={data} />
      </Table>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
