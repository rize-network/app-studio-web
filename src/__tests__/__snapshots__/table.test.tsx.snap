// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table to match snapshot 1`] = `
<table
  className="sc-bdfCDU fvLDyq"
  role="Table"
>
  <thead
    className="sc-bdfCDU gxlqWj"
  >
    <tr
      className="sc-bdfCDU cYBAia"
    >
      <th
        className="sc-bdfCDU frJuHz"
      >
        Invoice
      </th>
      <th
        className="sc-bdfCDU frJuHz"
      >
        Payment Status
      </th>
      <th
        className="sc-bdfCDU frJuHz"
      >
        Total Amount
      </th>
      <th
        className="sc-bdfCDU frJuHz"
      >
        Payment Method
      </th>
    </tr>
  </thead>
  <tbody
    className="sc-bdfCDU cYBAia"
  >
    <tr
      className="sc-bdfCDU cYBAia"
    >
      <td
        className="sc-bdfCDU jpUhUe"
        isFirstColumn={true}
      >
        INV001
      </td>
      <td
        className="sc-bdfCDU fPWZcZ"
        isFirstColumn={false}
      >
        Paid
      </td>
      <td
        className="sc-bdfCDU fPWZcZ"
        isFirstColumn={false}
      >
        $250.00
      </td>
      <td
        className="sc-bdfCDU fPWZcZ"
        isFirstColumn={false}
      >
        Credit Card
      </td>
    </tr>
  </tbody>
</table>
`;
